File: scoring/disagg_insights/writers/object_storage_writer.py

import logging

from pyspark.sql import DataFrame


class ObjectStorageWriter:
    logger = logging.getLogger(__name__)

    @staticmethod
    def write(output_df, output_path, mode):
        data_output_path = output_path + "/raw"
        if mode == "stream":
            query = (output_df
                     .writeStream
                     .format("csv")
                     .option("header", "true")
                     .option("checkpointLocation", output_path + "/_checkpoint")
                     .outputMode("append")
                     .start(data_output_path)
                     )
            query.awaitTermination()

        else:
            output_df.write.format("csv").option("header", "true").mode("append").save(data_output_path)

File: pre-process/src/main/scala/com/oracle/ugbu/ouai/ad/writers/ObjectStorageWriter.scala
package com.oracle.ugbu.ouai.ad.writers

import com.oracle.ugbu.ouai.ad.model.OutputRow
import org.apache.spark.sql.Dataset

object ObjectStorageWriter {

  def write(outputDF: Dataset[OutputRow], outputPath: String, mode: String): Unit = {
    val outputRawPath = outputPath + "/raw"
    if ("stream".equals(mode)) {
      val query = outputDF
        .writeStream
        .format("delta")
        .option("checkpointLocation", outputPath + "/_checkpoint")
        .outputMode("append")
        .start(outputRawPath)

      query.awaitTermination()
    }
    else {
      outputDF.write.format("delta").mode("append").save(outputRawPath)
    }
  }
}
